Test1/src/leetcode/LeetCode498.java

```diff
@@ -0,0 +1,44 @@
+package leetcode;
+
+/**
+ * https://leetcode.com/problems/diagonal-traverse/submissions/
+ * @author minibig
+ *
+ */
+public class LeetCode498 {
+
+	public int[] findDiagonalOrder(int[][] matrix) {
+		
+		int row = matrix.length;
+		int col = matrix[0].length;
+		
+		int arr[] = new int[row * col];
+		int idx = 0;
+		
+		for(int i = 0; i < row + col - 1; i++) {
+			
+			// moving down index
+			int r = i < col ? 0 : i - col + 1;
+			int c = i >= col ? col - 1 : i;
+			
+			// moving up index
+			int c2 = i < row ? 0 : i - row + 1;
+			int r2 = i >= row ? row - 1 : i;
+			
+			while(r < row && r >= 0 && c < col && c >= 0) {
+				if(i % 2 == 0) {
+					arr[idx++] = matrix[r2][c2];
+				}else {
+					arr[idx++] = matrix[r][c];
+				}
+				r++;
+				c--;
+				
+				r2--;
+				c2++;
+			}
```

Comment:
I think there's a better way to structure your code.

```
if (i % 2 == 0) {
	// moving up
	int r = i >= row ? row - 1 : i;
	int c = i < row ? 0 : i - row + 1;
	while(r < row && r >= 0 && c < col && c >= 0) {
		arr[idx++] = matrix[r][c];
		r--;
		c++;
	}
} else {
	// moving down
	int r = i < col ? 0 : i - col + 1;
	int c = i >= col ? col - 1 : i;
	while(r < row && r >= 0 && c < col && c >= 0) {
		arr[idx++] = matrix[r][c];
		r++;
		c--;
	}
}
```

Reply:
Another option:
```
int r, c, deltaR, deltaC;
if (i % 2 == 0) {
	// moving up
	r = i >= row ? row - 1 : i;
	c = i < row ? 0 : i - row + 1;
	deltaR = -1;
	deltaC = 1;
} else {
	// moving down
	r = i < col ? 0 : i - col + 1;
	c = i >= col ? col - 1 : i;
	deltaR = 1;
	deltaC = -1;
}
while(r < row && r >= 0 && c < col && c >= 0) {
	arr[idx++] = matrix[r][c];
	r += deltaR;
	c += deltaC;
}
```

I personally like the first implementation better, but you can also consider the second one.

The reason why the first implementation is better is that we can declare "r" and "c" with "final". Using "final" as much as possible is a good practice in Java.

You can read: https://stackoverflow.com/questions/154314/when-should-one-use-final-for-method-parameters-and-local-variables